DPMS and lock handling

snaggen opened the pull request:
This pull request adds DPMS and lock handling and requires some wlroots changes also, that is in a separate pullrequest for wlroots. 
This pullrequest have the following parts:
* idle notifications on mouse and keyhandling
* swayidle daemon that recieves idle notifications sends dpms commands and locks screen
* modifications to output command, adding dpms

Fixes #541

emersion commented on swaylock/main.c:
```diff
@@ -156,7 +157,30 @@ static const struct wl_registry_listener registry_listener = {
+}
+
 int main(int argc, char **argv) {
+    if (!create_file_lock()) {
```
I think we want to be able to start up multiple sway instances at the same time.

snaggen replied:
Well, it makes no sense to run mutilpe lockscreens in paralell, and it will fail anyway... so I thought it would be nice to handle this properly since this will be triggered if user starts swaylock and then swayidle starts another.

emersion commented on meson.build:
```diff
@@ -5,7 +5,7 @@ project(
 		'c_std=c11',
 		'warning_level=2',
-		'werror=true',
+		'werror=false',
```
We probably want to keep this to `true`.

snaggen replied:
Hmmm.. yes, that was not meant to slip in...

emersion commented on include/sway/server.h:
```diff
@@ -21,6 +21,7 @@ struct sway_server {
 
 	struct wlr_compositor *compositor;
 	struct wlr_data_device_manager *data_device_manager;
+    struct wlr_idle *idle;
```
We're using tabs for indentation.

snaggen commented:
Have updated this to use wlr_output_enable for DPMS, which means no modifications are needed in wlroots

snaggen commented:
Should have fixed all tabs vs spaces now...

emersion commented on include/sway/config.h:
```diff
@@ -98,6 +98,13 @@ struct seat_config {
 };
 
+enum config_dpms {
+    DPMS_IGNORE,
```
Not all actually :)

snaggen replied:
Well, yes... that is used too indicate that no DPMS configuration should be performed, if no dpms is set on the output command. And since enums defaults to the first entry, it is default.

emersion replied:
Oh, I was just talking about spaces vs tabs.

emersion commented on swayidle/meson.build:
```diff
@@ -0,0 +1,13 @@
+		wlroots,
+	],
+	link_with: [lib_sway_common, lib_sway_client],
+    install: true
```
You could grep for two spaces, or configure your editor to show invisible chars, or configure your editor to use the [`.editorconfig`](http://editorconfig.org/) file.

emersion replied:
This comment also indicates spaces instead of tabs.

snaggen commented:
Diff against master and grep for 2 spaces... fix! Next time I code, I will make sure to set vim to noet....

emersion commented on swayidle/main.c:
```diff
@@ -0,0 +1,207 @@
+
+static void handle_global(void *data, struct wl_registry *registry,
+		uint32_t name, const char *interface, uint32_t version) {
+	if (strcmp(interface, "org_kde_kwin_idle") == 0) {
```
A good practice is to use `org_kde_kwin_idle_interface.name` here.

emersion commented on swaylock/main.c:
```diff
@@ -156,7 +157,30 @@ static const struct wl_registry_listener registry_listener = {
+}
+
 int main(int argc, char **argv) {
+	if (!create_file_lock()) {
```
Bump: I still think this should be removed.

snaggen replied:
If you remove this, the only difference will be that you will get
zwlr_input_inhibit_manager_v1@8: error 0: this compositor already has input inhibited

Since the scenario where two swaylock might be run at the same time, will be a quite common one with the swayidle trying to run swaylock in the background. I thought it would be better to handle this in a more controlled fashion. 
Can you please argue, the pros of having swaylock exit by
zwlr_input_inhibit_manager_v1@8: error 0: this compositor already has input inhibited

emersion replied:
Why does this happen? `swayidle` shouldn't spawn more than one `swaylock` right?

ascent12 commented:
I have some problems with the overall design of this PR.

I really don't like that what is effectively a normal client (swayidle) is launching what should be a privileged client (a lockscreen). This just doesn't work from a security perspective.
A lockscreen really should be launched by the compositor itself in a special way, so that it can gain access to special protocols, although this type of filtering has not been implemented yet. Otherwise we're just going allow malicious "normal" clients to steal input and all sorts of nonsense like that, and basically bring back the problems X11 has.

I would personally prefer to see a DPMS option which would just do screen-blanking after so many minutes, just like it works for basically everything else. Having to `exec swayidle` in your config just seems extremely weird.

SirCmpwn commented:
>This just doesn't work from a security perspective

Note that this situation is something I think we can address with what I eventually have in mind for security stuff. Also, swayidle is going to be a privledged client at some point, when we add a protocol for DPMS so that the client can set that directly without using sway IPC (which is sway specific).

emersion commented:
Yeah, I'm not sure why there's an idle protocol in the first place.

ascent12 commented:
I don't see the value of this being a separate process. It's not like we're forking to keep privileged files open or anything like that.
To me, it just seems like a hack so you can use that KDE idle protocol as a client.

SirCmpwn commented:
The advantage is that the approach can be more complex without requiring every compositor to implement something themselves. With this approach you can:

- Lock the screen the screen after a timeout
- Launch a screensaver on top of that
- Dim the screen after a bit longer
- DPMS off after a bit longer
- Suspend the machine after a bit longer

If you really wanted to go all out, or any combination of these with any timings you prefer. All of it would (in theory) work on any compositor with the same idle daemon and no changes to the compositor or extra configuration.

ascent12 commented:
It's not like doing this kind of stuff is even a hard problem in the compositor itself.
```c
dim_source = wl_event_loop_add_timer(ev, dim_screen, data);
dpms_source = wl_event_loop_add_timer(ev, screen_off, data);
suspend_source = wl_event_loop_add_timer(ev, suspend, data);

// In input handling function
wl_event_source_timer_update(dim_source, 2 * 60 * 1000);
wl_event_source_timer_update(dpms_source, 5 * 60 * 1000);
wl_event_source_timer_update(suspend_source, 30 * 60 * 1000);
```
Or something extremely similar to that.

All this extra crap with another process and trying to be reusable is massively over-designing the solution.

SirCmpwn commented:
Well, yeah if you hardcode the timings and do those 3 specific things, it's easy as pie. But what if you don't?

ascent12 commented:
The numbers were just an example. It would obviously be something configurable with normal config options.

SirCmpwn commented:
Which is a nontrivial amount of code to write that every compositor is going to have to do and will probably do differently.

ascent12 commented:
I don't know why you're trying to write everybody else's compositors for them.
If it's really supposed to be shared like that, it should exist in wlroots.

SirCmpwn commented:
Aye, this is probably not the right place for it. I intend to file some follow-up tickets to generalize it a bit more, then pull it into a separate project along with swaybg and swaylock.

snaggen commented:
I must say that I'm a little split myself about having it in the compositor vs split out, having it in the compositor would make it work out of the box and it feels like it might be there as it is core functionality. But on the other hand, having it split out creates a clean separation and adds less clutter in the compositor code, as a compositor really should focus on compositing. In the end, I thing both ways are ok, and went with the daemon version due to being preferred by the lead dev (he should have better insight in to the design of the project). 
Also, about it feeling a little weird having to do exec swayidle in the sway config, this is how many other DE:s are doing this. Gnome have gnome-session starting the power-manager and other stuff. So this is really more a result of not having a sway-session, so that we need to do our own session manually by adding exec to the config.

snaggen commented:
Ok, so I have created https://github.com/swaywm/sway/pull/1831 and I actually think that is the preferred solution... but I'm not a manintainer, so it is your call.

snaggen commented:
Ported over the sleep handling and inhibiting from the other version to the daemon. Added a simple file signaling mechanism to swaylock to be able to determine if we have an active lockscreen.

snaggen commented:
Not a good idea to do what?

emersion commented:
>Added a simple file signaling mechanism to swaylock to be able to determine if we have an active lockscreen.

Let's say I have sway running on tty1 and another sway running on tty2. That means I can't lock both.

emersion commented:
>Not a good idea to do what?

To listen to the parent session logind DBUS API when we do not own the session.

snaggen commented:
So you mean that this should be moved in to the logind backend?

emersion commented:
>So you mean that this should be moved in to the logind backend?

When we said "this should be part of wlroots", that's what we meant yes. That's the right way to do it, but that also means external clients won't be able to listen to it directly and the compositor will have to handle it.

snaggen commented:
Ok

SirCmpwn commented:
In terms of "put it in sway/wlroots" versus "put it in a daemon", I think everyone has said their piece and I will go along with whichever solution @snaggen prefers.

snaggen commented:
After writing two versions that a functionally equivalent, it seems that the "in compositor" are not able to aquire the lock for some reason. Of course, we could spend more time digging in to the internals of systemd and dbus, but I really starting to lean towards that these kind of things shouldn't be inside the compositor. 
What all this is about is policy, and policy is better handled outside the compositor. With my latest changes to this daemon version I think I have handle most problems related to having it as a daemon. 
* The daemon is automatically started
* Configuration is handled as normal configuration
* Swayidle is a totally unprivileged daemon, requesting the compositor to do the privileged tasks. 
* The lock signaling file mechanism should allow to run multiple sway sessions in parallell, since it uses the wayland socket id in the filename. 

I consider this ready, so please review this and consider merging it.

snaggen commented on swayidle/main.c:
```diff
@@ -0,0 +1,399 @@
+		"sender='org.freedesktop.login1',"
+		"interface='org.freedesktop.login1.%s',"
+		"member='%s',"
+		"path='%s'";
```
Where should it be done do you mean? It is used the row right below, to construct the sd_bus_add_match parameter, following the style used in logind.c in wlroots.

emersion replied:
Ditto

snaggen commented on swayidle/main.c:
```diff
@@ -0,0 +1,399 @@
+	return 0;
+}
+
+void aquire_sleep_lock() {
```
why? This, the daemon version, does all its lock handling by it self... so it really need to aquire the sleep lock.

emersion replied:
We already talked about this, the systemd part should be in wlroots.

somdoron replied:
@emersion acquiring the lock with in the compositor doesn't work. This is way the daemon solution.

emersion replied:
There's no reason acquiring the lock in the compositor wouldn't work. Debug it and fix it?

The daemon doesn't have ownership over the session, it shouldn't lock a parent session if the compositor isn't using DRM+logind.

snaggen commented on swaylock/main.c:
```diff
@@ -156,7 +157,64 @@ static const struct wl_registry_listener registry_listener = {
 
 static struct swaylock_state state;
 
+bool create_file_lock() {
```
Why? If the user starts a lock screen, and leaves the sceen, then swayidle will try to start a lockscreen and get an error. I prefer handling known race scenarious and exit cleanly. I will update this to include the WAYLAND_DISPLAY in the path, so you can run lockscreens in paralell sway instances.

emersion replied:
Including `WAYLAND_DISPLAY` is a little bit better.

Why not having the compositor do this test? The compositor knows if the swaylock is running, because it started it.

SirCmpwn replied:
The compositor will use the already_inhibited error on input-inhibitor, which aborts any swaylock clients after the first.

snaggen commented on sway/main.c:
```diff
@@ -413,7 +439,7 @@ int main(int argc, char **argv) {
 	// TODO: consume config->cmd_queue
 	config->active = true;
-
+	invoke_swayidle();
```
One of the conserns about having this in a daemon would be that it is weird that the user would have do do config to get this kind of basic functionality. And I agree. This kind of things should be run as default, with a possible option to opt out. 
And there are actually a few gains of running it as a daemon even if it is always started. 
* It runs in parallel, no blocking of the compositor. I accidentally added a busy wait loop in my lockscreen check, and froze the compositor for a second. And even with no stupid misstake like that, you are still performing non-compositing work in the compositing mainloop. 
* Crash safety, even if a bug appears in the idle code, or a third party lib like sd_bus causing  a crash, you still only crash the idle manager and not the compositor. 
* Code separation, non-compositing related code are separated out to its own daemon. 

However, I understand your point, so I'll add a check and if both idle and lock timeouts are zero, no daemon is started. That way you get an opt out.

emersion replied:
The main advantage of having an external daemon is to be able to replace it with another one.

>This kind of things should be run as default, with a possible option to opt out.

You can put it in the default config. No need to hardcode it anywhere.

snaggen commented on include/sway/config.h:
```diff
@@ -305,6 +313,10 @@ struct sway_config {
 	char *font;
 	int font_height;
+	uint32_t idle_timeout;
+	uint32_t lock_timeout;
```
Adding opt out if both are zero. having the user to setup a swayidle_command for basic functionallity makes no sense. This functionality should be there by default, if someone wants to opt out, the can do that by setting them to zero.

SirCmpwn replied:
Nah, we should make it opt-in. Having to opt out of DPMS always annoyed me.

emersion commented:
Maybe the "systemd is in the daemon" part can be file as a follow-up issue.

emersion commented on swayidle/main.c:
```diff
@@ -0,0 +1,397 @@
+	wlr_log(L_DEBUG, "Checking for lock");
+	int inhibit_cnt = 0;
+	while(!have_lock() && inhibit_cnt < 5) {
+		nanosleep(&tv, NULL);
```
Polling is evil.

snaggen replied:
Why? This is in its own thread, so if it wasn't sleeping it would still not do anything. This is for all practical reasons completely equivalent of having a timer on a eventloop, since this thread doesn't have any other thing to do. The difference in this case would be absolutely zero.

SirCmpwn replied:
>Why? This is in its own thread

Oh god why does this have threads

emersion replied:
Because this is just a huge, ugly hack. This isn't how you "wait for an event be be triggered".

snaggen replied:
sd_bus_wait will block waiting for the next request, so it felt easier to just have a simple thread than to start up a full eventloop just for this...

SirCmpwn replied:
Threads are the devil. Can't you somehow get the underlying dbus fd and poll (as in poll.h) on it?

SirCmpwn commented:
This is looking okay, but it needs a lot of work.

- Review the style guide and fix the errors
- swayidle needs a man page
- swayidle should use an event loop instead of threads. Use the wayland-server event loop if you want,  we already depend on it
- Instead of adding specific sway commands like this, I imagined a different approach, something like this:

```
$ swayidle \
    logind-sleep exec "swaymsg exec swaylock"
    timeout 10m exec "swaymsg exec swaylock" \
    timeout 20m dpms on
```

Then instead of having sway know much about it the user should just `exec` it themselves.

snaggen commented:
Have a link to the style guide? 
Think I have adressed most other things, like the eventloop, letting compositor find the lock, let already_inhibited handle multiple lockscreens. Removed the hardcoded autostart, just adding it to the default config, removed the idle/lock_timeout and adding a man-page.

snaggen commented:
Focusing on moving the systemd parts to the eventloop I forgot the other parts... so this is currently most likely not working I will fix this and go over the style (if I find the style guide) later.

snaggen commented:
Ok, it seems I need some assistance from the people who actually know wayland here.... 
For the event loop i used `wl_event_loop_create()` to create an event loop, and run that with a
```
while (wl_event_loop_dispatch(state.event_loop, 0) != -1) {
; //Intentionally left blank;
}

```
Now I need to hook up the wl_display_dispatch to that loop, and my assumption was that I should do something like 
```
wl_event_loop_add_fd(state.event_loop, wl_display_get_fd(state.display),
            WL_EVENT_READABLE, display_event, NULL);

```
and in the function display_event call 
`wl_display_dispatch(state.display);`
However, this doesn't seem to work... so Im a little lost here... so any input would be appreciated.

SirCmpwn commented:
If it's easier you could use the swaybar event loop, which you can reference to get things working.

snaggen commented:
Ok, I'll try that.

snaggen commented:
Now it is working again, I also did some minor clean up and looked over the code to fix obvious style issues.

snaggen commented:
I added the possibility to run arbitrary commands on idle, I support forking and executing commands in a shell, and sending commands to sway. However, to keep the "just works" part, it will by default setup a dpms and lock listener, you can control the timeout of these using the -i and -l (as before). Setting the to 0 will disable them. If you want to disable them and set them up yourself, you can run:

> swayidle -i 0 -l 0 timeout 600 command "output * dpms off" resume exec "output * dpms on" \
>                                            timeout 300 command "lock"

SirCmpwn commented:
I don't want the just works behavior, it's no big deal to make people be explicit about it.

snaggen commented:
Why? Why is it bad that things works, without people need to learn every knob in every daemon to get a sane behaviour? I have added the possibility to disable it all, for the people willing to learn this...

snaggen commented:
However, the systemd sleep lock is not possible to setup by commandline, since that includes checking and releasing the lock. You can control that by specifying -x on the command line to disable it.

SirCmpwn commented:
>Why? Why is it bad that things works, without people need to learn every knob in every daemon to get a sane behaviour? I have added the possibility to disable it all, for the people willing to learn this...

Just gonna go with "because I said so". It's a matter of personal preference.

>However, the systemd sleep lock is not possible to setup by commandline, since that includes checking and releasing the lock. You can control that by specifying -x on the command line to disable it.

Invert it, pass a flag to enable it

With no command line args swayidle should do nothing (except maybe print a warning telling you it's going to do nothing).

SirCmpwn commented on swayidle/main.c:
```diff
@@ -0,0 +1,481 @@
+	bool locked = json_object_get_boolean(is_active_json);
+	wlr_log(L_DEBUG,"Got lock? %d", locked);
+	return locked;
+}
```
What do you need to know this for?

snaggen replied:
That is to know when the lockscreen have been raised, so I can release the sleep lock.

SirCmpwn replied:
I suppose I'm not clear on what the sleep lock is supposed to do.

snaggen replied:
The sleep lock prevents the laptop from going to sleep until it is release. We use this to ensure that the lockscreen is up before the system is going to sleep. If we do not release the lock, the system will have a delay on about 5 seconds before going to sleep. 
If we do not use a sleep lock, the system may go to sleep before the lock is up. This will cause any person that opens the laptop to see the content on the screen on resume, before the lockscreen is up.

SirCmpwn replied:
I see. Rather than an IPC command can't we just assume that if we ran swaylock the lock screen is up?

snaggen commented:
I guess "Because I said so" is a reason as goo as any :) 
I'll fix this later today. I'll just remove the default, but keep the -i -l options convenience options. 
If nothing is specified, no timers are created, so I'll make swayidle just exit then.

SirCmpwn commented on config.in:
```diff
@@ -5,7 +5,7 @@
 
 #Session setup
-exec swayidle
+exec swayidle -i 600 -l 300 -x
```
Take this out of the default config, or at least comment it by default. If you choose the latter extend the comment to explain what these args do and suggest the man page for further reading.

snaggen replied:
So, there shoud not be any dpms/lockscreen handling by default?

SirCmpwn replied:
Correct

SirCmpwn commented on sway/meson.build:
```diff
@@ -58,6 +58,8 @@ sway_sources = files(
 	'commands/workspace.c',
 	'commands/ws_auto_back_and_forth.c',
+	'commands/lock.c',
+	'commands/swaylock_command.c',
```
These should be taken out, and swayidle shouldn't use sway IPC directly. It should just let the user execute commands on a timeout, such as `swaymsg exec swaylock ...`

snaggen replied:
Well I understand that... but I really think that it would be nice for the user to have a consistent way to run the lockscreen, so it looks the same everytime. And for the systemd handling that executes the lockscreen, it is impossible to know how the user wants it. So it makes sense have a sway command for that, so that it is possible to just let sway lock the screen and get the same result everytime. This also have the benefit that the user can bind a key to the lock command and get the same lockscreen as swayidle uses for the systemd part.

SirCmpwn replied:
>I really think that it would be nice for the user to have a consistent way to run the lockscreen, so it looks the same everytime

Just run it with the same arguments every time, then. If you want to run it from several places with the same arguments without having to update all of them every time you change it, wrap swaylock in a script.

snaggen replied:
Well, since the sleep lock part, is nothing we get from the user (since that will make it impossible to know how to verify it is ok to release the lock), it is not possible to get this from the user. This is the reason I added the lock and swaylock_command.

SirCmpwn replied:
What is the "sleep lock" part?

snaggen replied:
Is there a problem having a lock and swaylock_command? Or is there a problem with swayidle doing ipc? I'm not really sure I understand the issues caused by this?

SirCmpwn replied:
There is, yes. I intend to split swaylock out into a standalone project that works on many compositors. Users might want to lock with a different lock screen as well. It also adds unnecessary complexity, it's no better than just having swayidle take an arbitrary command.

snaggen replied:
I assume that if you can help me make the IPC_IS_LOCKSCREEN_ACTIVE more generic (currently it looks at the name of the lockscreen surface), I guess we could just let the -x parameter take the lock command to use.

SirCmpwn replied:
I don't think IPC has to be involved here at all. A parameter specifying the lockscreen would be good but not sure why swayidle has to know the command it's running is a lockscreen at all

SirCmpwn commented:
Still seeing style errors, please review the style guide

snaggen commented:
I haven't found the style guide and I haven't gotten a link either...

SirCmpwn commented:
https://github.com/swaywm/sway/blob/master/CONTRIBUTING.md

snaggen commented:
Have look at the style guilde and reviewed the code.

SirCmpwn commented on swayidle/main.c:
```diff
@@ -0,0 +1,497 @@
+	if (data == NULL) {
+		return;
+	}
+	char *param = (char *) data;
```
style: no space after )

SirCmpwn commented on swayidle/main.c:
```diff
@@ -0,0 +1,497 @@
+
+static int release_lock(void *data) {
+	wlr_log(L_INFO, "Releasing sleep lock %d", ongoing_fd);
+	if (ongoing_fd>=0) {
```
style: spaces around operators

SirCmpwn commented on swayidle/main.c:
```diff
@@ -0,0 +1,497 @@
+	}
+
+	inhibit_cnt++;
+	if(inhibit_cnt > 10) {
```
style: space before (

SirCmpwn commented on swayidle/main.c:
```diff
@@ -0,0 +1,497 @@
+	} else {
+		ret = sd_bus_message_read(msg, "h", &lock_fd);
+		if (ret < 0) {
+			wlr_log(L_ERROR, "Failed to parse D-Bus response for Inhibit: %s", strerror(-ret));
```
style: 80 column limit

SirCmpwn commented:
We should probably think about how to incorporate idle inhibit into this. Updating the `idle` protocol to specify whether idle inhibit is active would probably be good.

emersion commented:
>Updating the idle protocol to specify whether idle inhibit is active would probably be good.

No need for this: just prevent the timers from being triggered while idle inhibit is active?

SirCmpwn commented:
You can just assume that the command is a lock screen, or rather is whatever the user wanted you to start (maybe not a lockscreen). You can wait a second or something if you want.

snaggen commented:
Well, just waiting a second and hoping for the best is not really a good design choise ... especially for a security related function.

snaggen commented:
But if that is your preferred solution...

SirCmpwn commented:
Well, realistically lay out the worst case scenario for me.

snaggen commented:
Looked through the git diff upstream/master output and not it seems that it only contains the things needed.

snaggen commented:
Rebase done.

snaggen commented:
Updated the spelling pointed out... thanks for the review, not native english speaker, so its nice to get help with this.

snaggen commented:
Updated the manpage, removed the comment about the default behaviour. Also, removed the socket options, since that was removed

aenda commented on protocols/idle.xml:
```diff
@@ -0,0 +1,49 @@
+  ]]></copyright>
+  <interface  name="org_kde_kwin_idle" version="1">
+      <description summary="User idle time manager">
+        This interface allows to monitor user idle time on a given seat. The interface
```
This is a grammar nitpick, but "This interface allows for the monitoring of user idle time on a given seat." is probably better.

SirCmpwn replied:
This isn't our protocol, take it up with KDE

aenda commented on protocols/idle.xml:
```diff
@@ -0,0 +1,49 @@
+  <interface  name="org_kde_kwin_idle" version="1">
+      <description summary="User idle time manager">
+        This interface allows to monitor user idle time on a given seat. The interface
+        allows to register timers which trigger after no user activity was registered
```
Similarly, "allows the registration of timers which trigger..." (going off of https://english.stackexchange.com/questions/85069)

aenda commented on sway/config/output.c:
```diff
@@ -204,6 +207,20 @@ void apply_output_config(struct output_config *oc, struct sway_container *output
+	if (oc && oc->dpms_state != DPMS_IGNORE) {
+		switch(oc->dpms_state) {
+			case DPMS_ON:
+				wlr_log(L_DEBUG, "Turning on scren");
```
"Turning on screen"

aenda commented:
I added some suggestions for spelling/grammar fixes as comments on the PR (given my lack of C experience, that's the most helpful I think I can be). Looking forward to this functionality getting implemented!

aenda commented on swayidle/swayidle.1.txt:
```diff
@@ -0,0 +1,59 @@
+'swayidle' [options] [timeout <seconds> exec <parameter>
+			[resume exec <parameter>]] ...
+
+Swayidle handles idle events from sway, and setup lockscreen and turns off
```
"from sway: it sets up the lockscreen and turns off the screen." might be better, but this isn't as clear cut

snaggen commented:
Updated the spelling and wording in the manpage. Also updated the branch against latest master and rebased, for easy merging.

snaggen commented:
I noticed that this pull request haven't been merged yet, which I assume is due to much work and little time... but, I also want to make sure I haven't missed anything, so you are waiting for me to do something...  since I live under the impression that I have fixed all the requested changes, and that this is ready for merge.

SirCmpwn commented:
Sorry, I've just been busy and this is a big patch. I'll review it again soon.

snaggen commented:
No, problem... just worried that you were waiting for me, without me realizing it...

SirCmpwn commented on sway/config/output.c:
```diff
@@ -204,6 +207,20 @@ void apply_output_config(struct output_config *oc, struct sway_container *output
 	}
+	if (oc && oc->dpms_state != DPMS_IGNORE) {
+		switch(oc->dpms_state) {
+			case DPMS_ON:
```
Style: deindent the case within the switch and add a space so `switch(` becomes `switch (`

snaggen replied:
Done

SirCmpwn commented on swayidle/main.c:
```diff
@@ -0,0 +1,448 @@
+int parse_timeout(int argc, char **argv) {
+	if (argc < 4) {
+		wlr_log(L_ERROR, "Too few parameters to timeout command. \
+Usage: timeout <seconds> exec <command>");
```
Prefer multiple strings to backslashes

```c
    wlr_log(L_ERROR, "Too few parameters to timeout command. "
        "Usage: timeout <seconds> exec <command>");
```

snaggen replied:
Done

SirCmpwn commented on swayidle/main.c:
```diff
@@ -0,0 +1,448 @@
+	}
+	struct swayidle_timeout_cmd *cmd =
+		calloc(1, sizeof(struct swayidle_timeout_cmd));
+	cmd->timeout = seconds*1000;
```
Put spaces around operators

snaggen replied:
done

SirCmpwn commented on swayidle/main.c:
```diff
@@ -0,0 +1,448 @@
+			printf("Usage: %s [OPTIONS]\n", argv[0]);
+			printf("  -i seconds optional\t\tidle timeout in seconds\n");
+			printf("  -l seconds optional\t\tlock timeout in seconds\n");
+			printf("  -x lock_command\t\t to raise the lockscreen on systemd \
```
This looks weird in my terminal, can you align with spaces instead

snaggen replied:
This is fixed

SirCmpwn commented on swayidle/main.c:
```diff
@@ -0,0 +1,448 @@
+			exit(-1);
+		}
+	}
+	return 0;
```
If the user doesn't configure anything you should probably warn them with some logging

snaggen replied:
That is done at: 
    if (!should_run) {
        wlr_log(L_INFO, "No command specified! Nothing to do, will exit");
        sway_terminate(0);
    }

SirCmpwn commented:
This is looking much better, nice work.

SirCmpwn commented:
The `-l` and `-l` options should be removed and users should instead just use the normal command line syntax.

snaggen commented:
I thought it would be nice to have convenience options for the most common use cases. But, I'll remove them if you don't like it.

SirCmpwn commented:
Yeah, I'm not a huge fan of those.

SirCmpwn commented:
Also, with the approach refined a bit now I'd be open to having it (commented out) in the default config again.

SirCmpwn commented on swayidle/swayidle.1.txt:
```diff
@@ -0,0 +1,59 @@
+	number of seconds.
+
+*-x <lock command>*::
+	Raise the lockscreen using the lock command, upon systemd sleep events.
```
This should be documented as executed with a shell (and if it's not executed with a shell, it should be).

SirCmpwn commented on config.in:
```diff
@@ -4,6 +4,11 @@
+### Session setup
+#
+# Power off screen on idle and raise lockscreen
+exec swayidle -x 'swaylock -c 000000' timeout 300 exec 'swaylock -c 000000' timeout 600 exec 'swaymsg "output * dpms off"' resume exec 'swaymsg "output * dpms on"'
```
Can you split this onto several lines with \ and comment it out by default?

snaggen replied:
Done

SirCmpwn commented on swayidle/swayidle.1.txt:
```diff
@@ -36,20 +36,19 @@ Options
+swayidle -x \'swaylock -c 000000' timeout 300 exec \'swaylock -c 000000' 
+timeout 600 exec \'swaymsg "output * dpms off"' resume exec \'swaymsg "output *
+dpms on"'
+
```
Let's not try to maintain two examples, just stick with the one in the default config

SirCmpwn commented on swayidle/swayidle.1.txt:
```diff
@@ -36,20 +36,19 @@ Options
-
 *-x <lock command>*::
-	Raise the lockscreen using the lock command, upon systemd sleep events.
+	Raise the lockscreen using the lock command (by executing it in a shell), upon systemd sleep events.
```
Hmm.

Let's get away from this being a flag and get rid of references to the lockscreen. Let's make it behave like timeout/resume:

```
swayidle before-sleep exec 'swaylock...'
```

snaggen replied:
done

SirCmpwn commented on config.in:
```diff
@@ -4,6 +4,13 @@
+# Power off screen on idle and raise lockscreen
+#exec swayidle -x 'swaylock -c 000000' \
+#	timeout 300 exec 'swaylock -c 000000' \
+#	timeout 600 exec 'swaymsg "output * dpms off"' resume exec 'swaymsg "output * dpms on"'
```
Can put another newline for resume

snaggen replied:
done

SirCmpwn commented on swayidle/swayidle.1.txt:
```diff
@@ -0,0 +1,51 @@
+[timeout <seconds> exec <parameter> [resume exec
+<parameter>]]
+
+It is possible to pass multiple commands to setup multiple idle handlers.
```
This could be better at describing each command line "subcommand". I can take a crack at rewriting it if you like

snaggen replied:
Its probably better, since my native tongue is Swedish...

snaggen commented:
Rebased against master... going to sleep now. Let me know if there is more work needed.

SirCmpwn commented:
I pushed a new man page, let me know what you think.

One more change, let's get rid of `timeout <n> exec <cmd>` and just make it `timeout <n> <cmd>`. Already updated the man page assuming this.

Thanks for your patience and hard work, this is looking really great!

snaggen commented:
Thanks for the help with the manpage, it looks great! I have updated swayidle to remove the redundant exec command, as in the manpage. I also noticed that one extra exec was missed in the default config for the before-sleep command, so I fixed that also.

snaggen commented:
Realized I had missed some review comments above... but now I have fixed all issues mentioned, so this should be ready to merge.

SirCmpwn commented:
Beautiful. Great work.

emersion commented on sway/commands/output.c:
```diff
@@ -20,6 +20,22 @@ static char *bg_options[] = {
+		output->dpms_state = DPMS_ON;
+	} else if (strcmp(value, "off") == 0) {
+		output->dpms_state = DPMS_OFF;
+	}
```
Can you return an error if it's neither "on" nor "off"?

snaggen replied:
Done

emersion commented on swayidle/main.c:
```diff
@@ -0,0 +1,414 @@
+
+int main(int argc, char *argv[]) {
+	signal(SIGINT, sig_handler);
+	signal(SIGTERM, sig_handler);
```
Nitpick: is this really needed?

snaggen replied:
Well, I guess it isn't striclty needed... but I think it is good to handle SIGINT and SIGTERM when you have a long running process, that handles shutdown in a controlled way.

emersion replied:
If the handler is just calling `exit(0)`, imho it can be removed

snaggen replied:
Well, it destroys the eventloop and disconnects from the display... so it is not just exit 0... not sure how important it is to destroy the eventloop nicely and disconnect from the display... but, I don't see any reason not to do it.

emersion replied:
Ah, you're right, `sway_terminate` is not _just_ `exit(0)`.

emersion commented on swayidle/swayidle.1.txt:
```diff
@@ -0,0 +1,67 @@
+/////
```
Follow-up issue: rewrite this in scdoc

snaggen replied:
Done.

SirCmpwn replied:
Remove the asciidoc file too

emersion replied:
Bump

snaggen replied:
asciidoc file removed...

emersion commented:
Oh, and feel free to come on join us on our IRC channel (#sway-devel on Freenode).

snaggen commented:
Ok, have fixed these last comments, and have rebased against master resolving the scd conflicts... so now it should be merge:able again.

SirCmpwn commented on swayidle/swayidle.1.scd:
```diff
@@ -0,0 +1,59 @@
+
+Maintained by Drew DeVault <sir@cmpwn.com>, who is assisted by other open
+source contributors. For more information about sway development, see
+<https://github.com/swaywm/sway>.
```
No <> around the link here

snaggen replied:
done

SirCmpwn commented on swayidle/swayidle.1.scd:
```diff
@@ -0,0 +1,59 @@
+     timeout 300 'swaylock -c 000000' \
+     timeout 600 'swaymsg "output * dpms off"' \
+		resume 'swaymsg "output * dpms on"' \
+     before-sleep 'swaylock -c 000000'
```
This should be in a literal block, use \`\`\` (see man scdoc)

snaggen replied:
Done

SirCmpwn commented on swayidle/swayidle.1.scd:
```diff
@@ -0,0 +1,59 @@
+
+*timeout* <timeout> <timeout command> [resume <resume command>]::
+	Execute <timeout command> if there is no activity for <timeout> seconds.
+	+
```
No + here

snaggen replied:
Done

SirCmpwn commented on swayidle/swayidle.1.scd:
```diff
@@ -0,0 +1,59 @@
+
+#EVENTS
+
+*timeout* <timeout> <timeout command> [resume <resume command>]::
```
No :: here

snaggen replied:
Done

SirCmpwn commented on swayidle/swayidle.1.scd:
```diff
@@ -0,0 +1,59 @@
+on various idle-related events. You can specify any number of events at the
+command line.
+
+#EVENTS
```
Space after \#

emersion commented on swayidle/swayidle.1.scd:
```diff
@@ -0,0 +1,59 @@
+#EVENTS
+
+*timeout* <timeout> <timeout command> [resume <resume command>]::
+	Execute <timeout command> if there is no activity for <timeout> seconds.
```
Also `_` should be used instead of `<` and `>` in the description: `Execute _timeout command_ if …`

snaggen replied:
Done

SirCmpwn commented:
scdoc is very strict about formatting - did you actually run it on your changes?

SirCmpwn commented:
Ah, so it goes: https://builds.sr.ht/job/3079#task-build-14

snaggen commented:
Hmmm... seems I don't have scdoc installed, and that is probably why the build succeeds anyway...

snaggen commented:
Ok, I fixed the scdoc comments above, and the checks passes... so can you please see if the sayidle manpage looks ok now? I fail to find a scdoc package for fedora.... seems I have lost my googling mojo.

snaggen commented:
built scdoc from source and were able to test the scd version of the man page. Fixed a minor issue and rebased to squash the small fixes.

snaggen commented:
Sorry for this unexpected hassle, due to me lacking scdoc... but now it should be ok again and ready to merge.

SirCmpwn commented:
Thanks!
